src/ai/backend/gateway/vfolder.py

```diff
@@ -910,6 +913,74 @@ def _do_ops():
     return web.json_response(resp, status=200)
 
 
+async def download_directory_as_archive(request: web.Request,
+                                        file_path: Path,
+                                        zip_filename: str = None) -> web.StreamResponse:
+    """Serve a directory as a zip archive on the fly."""
+    class _Sentinel:
+        pass
+
+    _sentinel = _Sentinel()
+
+    def _iter2aiter(iter):
+        """Iterable to async iterable"""
+        def _consume(loop, iter, q):
+            for item in iter:
+                while True:
+                    if not q.full():
+                        q.put(item)
+                        break
+                    time.sleep(0.01)
```

Comment:
Just use `q.put(item)` and it will block for you when the queue is full and resume the operation automatically when the consumer makes a free slot. While blocked, `iter` is also paused. Your pattern is only required for `put_nowait()` and currently I don't see any reason to stick with that.
We are already doing the flow-control by limiting the maximum queue size.

Reply:
Reduced above code block with just `q.put(item)`.